### ISSUE TITLE
Trigger job with custom parameters

### ISSUE THREAD

**Issue by cameron-r** [concourse/concourse#783]
# Feature Request

## What challenge are you facing?

I want to build a pipeline that has a deploy step that can deploy to many different environments, eg:

build image -> test image -> deploy

where the 'deploy' job can deploy to one of several environments. We have a bunch of servers running that different teams use (around 20). Currently, it seems like with concourse, to be able to quickly deploy to any one of those 20 environments, we need to create 20 separate deploy jobs, which is not the ideal (especially when I think about having pipelines for multiple microservices as we build them out).

## A Modest Proposal

The ideal for us would be if when someone kicks off the 'deploy' job, there would be a UI prompt for them to enter what environment they want (eg 'environment-12') and that gets passed to the appropriate task

**Comment by cameron-r** [comment]
Yup, I definitely agree that allowing users to put custom parameters into builds could facilitate a lot of not-so-great patterns. It is, however, the existing 'common' solution in other CI/CD tools for the problem I want to solve with Concourse.

Just addressing your suggested work-arounds:

**Deploying to all environments:** We don't want to deploy to multiple environments at the same time, we kind of want them to be isolated from each other for now (there are multiple teams working on different projects, and we don't want project A's changes to disrupt project B)

**Write 20 jobs:** 20 jobs is not ideal since maintaining them would be more difficult (either by hand, or by creating some custom script to generate the jobs). Even from a UI perspective, I don't want to find my one job out of 20 each time, or clicking through each of them when I only care about who deployed in the past 2 hours (or something).

Maybe there is a solution in having a specific option you can give to Resources that tells them to prompt the user for input if there's a Put in that step. I'm trying to think of something that solves my problem, but still constrains users enough to discourage using it for bad patterns.

**Comment by brainsik** [comment]
I have a similar use case I'm trying to solve. A client of mine uses Jenkins to deploy new development environments. Those environments are created via AWS CloudFormation stacks which take a username parameter that's used to set a DNS record pointing at the host which will be created. These are specifically deployed ad-hoc / on-demand so people can blow away and get new environments whenever they want / need them. There are quite a few people at this company, so a pipeline per person is not a great option.

One workaround that seems possible is to have a task with params and then use `fly` to run the task with the parameter set. This feels pretty hacky though as it's not tracked in a visible way. Nothing is in the UI and running `fly builds` just shows them as "one-off" without any other detail.

Perhaps a way to alleviate some of @k0s's concerns would be if these jobs could only exist as end nodes in the graph. Meaning, nothing else would triggered, output, etc. from them.

**Comment by holycheater** [comment]
Yeah. That would be nice. Because if I have for example a set of pull requests as resources. I build&test each pull-request in a job. And then I have another job where I optionally want to deploy specific version (specific pull-request) to some staging area.

**Comment by chendrix** [comment]
Have you tried using the [pool-resource](github.com/concourse/pool-resource) to model your environments?

Alternatively can't you use [input_mapping](http://concourse.ci/task-step.html#input_mapping) to use the same task step but specify a different environment to deploy to?

**Comment by tmszdmsk** [comment]
We have a similar use case. We are considering moving from Jenkins to concourse but we use parametrised builds in dev related jobs (e.g. deploy this branch on this dev env). I think that having a resource that can be triggered from web ui/fly tool with a set of predefined parameters that will be stored on `out` as files would be great.

**Comment by chendrix** [comment]
Looks like something also desired in #903

**Comment by clarafu** [comment]
@cameron-r I'm also wondering if you can commit the environment name to a git resource and have the deploy job only triggered off that git resource and read the environment name from it. That way you can keep a history of which environments have been triggered and it will have a somewhat similar flow to a UI prompt.

**Comment by pieterza** [comment]
Any update on this?  We also have this requirement.

**Comment by dadacafe** [comment]
+1 for this request. We also have a need for such feature.

**Comment by mustafaakin** [comment]
We have this requirement also and willing to contribute

**Comment by ringods** [comment]
One of the redesigns in progress is around [Resources v2](https://github.com/concourse/rfcs/pull/1). Part of this proposal is resources spaces, originally tracked in concourse/concourse#1707. Spaces are the generic solution to support matrix builds.

I was thinking that with a proper design, manual input could be one of the different spaces which could be supported. Just tying one idea to another... 😃

**Comment by res0nat0r** [comment]
I'm essentially looking to do the same. Right now we use Jenkins parameterized builds to let our developers use our large Rake code to launch new environments via chef like:

`bundle exec rake "launch_environment[web,prod,101]"`

I would like to be able to replicate this in concourse by passing the rake inputs to a pipeline via trigger-job, something like:

`fly -t concourse trigger-job ops/launch_environment -v COOKBOOK=web -v ENVIRONMENT=prod -v ENV_NUM=101`

**Comment by spmsaj** [comment]
curious, is this a possibility ( ability to trigger job and pass new variables as part of it?)

**Comment by LunaGao** [comment]
I need this too.

**Comment by jeffreylim17** [comment]
Any updates on this one?

**Comment by luwang-vmware** [comment]
Concourse is a brilliant CI and it has easy simple interface to create a pipeline for testing. I like it.
In our daily work, I realize that supporting custom parameter in build triggering is very important for us.  It could give developer/tester more flexibility.  Supporting custom parameter with GUI or fly cli are both ok. Hope to see it soon.

**Comment by geofffranks** [comment]
Still looking forward to this 🤞

**Comment by dfuentes77** [comment]
This is really needed

**Comment by dfuentes77** [comment]
> Have you tried using the [pool-resource](github.com/concourse/pool-resource) to model your environments?
> 
> Alternatively can't you use [input_mapping](http://concourse.ci/task-step.html#input_mapping) to use the same task step but specify a different environment to deploy to?
> 
> The concourse team (and many many Cloud Foundry teams) uses the pool resource when choosing one of several different environments to deploy to.

Can anyone show an example of this for the purposes of deploying to multiple environments?

**Comment by freelock** [comment]
Hi,

Like other commenters, we would find this useful for provisioning development copies for individual developer use. 

As a workaround, I've been thinking of using the Matrix Notification Resource for this -- right now it's mainly used to push messages into Matrix from Concourse from build jobs, but Matrix does allow storing state in rooms, and our bot already manages this state as well as triggers Concourse jobs. We could set it up so that Concourse could read state from a Matrix room, as a resource to load info inside a job -

However, we currently use the Matrix Notification resource only in a single Matrix room as a sort of "message bus" -- our jobs post results there, and our bot relays the data and messages to project-specific rooms. So it would be quite a bit simpler for us to have our bot pass a parameter to the Trigger Job it already does, than to add the Matrix resource to a bunch of rooms and have it read/write state directly...

**Comment by dfuentes77** [comment]
I wholeheartedly agree.  The simplicity of just passing a parameter in the UI to a job and have it run, for individual and unique PR based deploys, would be extremely beneficial to us.

**Comment by geofffranks** [comment]
+1 again

**Comment by pivotal-jamie-klassen** [comment]
[instanced pipelines](https://github.com/concourse/rfcs/pull/34) is a pretty interesting abstraction that could, with the right UX, cover the use-cases described here pretty nicely.

**Comment by geofffranks** [comment]
+1

**Comment by dfuentes77** [comment]
+1

**Comment by Xercoy** [comment]
+1

**Comment by mkrish004c** [comment]
+1

**Comment by v3nturetheworld** [comment]
+1

**Comment by dhilgaertner** [comment]
+1

**Comment by drewhemm** [comment]
A CI system that doesn't support parameterised builds!? We are currently running a POC for migrating to Concourse from Jenkins and I don't want to have to create _n_ versions of a job to cater for _n_ parameter permutations...

**Comment by drewhemm** [comment]
For additional context, a task runner can be used for many other things than compiling source code and packaging applications. There are many use cases for just doing _x_, and _x_ needs to be doable in different environments...

**Comment by pivotal-jamie-klassen** [comment]
@drewhemm if you don't need the full auditability of "doing _x_" being linked to to an immutable history of versions (which you might not, if you're not compiling source or packaging applications), you could use [`fly execute`](https://concourse-ci.org/tasks.html#running-tasks). Especially true if the subtext of "_just_ doing _x_" could meaningfully include "... as a one-off", which it might given your use of the more-generic term "task runner".

Concourse is pretty opinionated about being stateless _and_ auditable, and it's tricky to satisfy these constraints while allowing parameterised builds in the way that most people are familiar with. However, feedback similar to yours has been flowing in for a long time, so it's a goal that is being steadily worked towards, as discussed pretty thoroughly in [this blog post](https://blog.concourse-ci.org/core-roadmap-towards-v10/#rfc-29-spatial-resources) and [this RFC](https://github.com/concourse/rfcs/pull/29).

Jenkins vs Concourse is always an interesting comparison, but based on your interrobang I get the sense that you're not doing it for 'basic research' purposes. I don't think Concourse has an explicit mandate to compete with Jenkins; the niche it fits is a bit more about people who are planning to implement a few strongly held opinions about automation. Those opinions tend not to be held quite as strongly in other CI systems, so Concourse winds up distinguishing itself in the CI use case.

**Comment by drewhemm** [comment]
Thanks @jamieklassen I'll definitely take a look at `fly execute`.  

All we need really is the ability to run tasks, sometimes manually with parameters and record those parameters entered and the log of the job; this is for auditing of actions performed against an environment, primarily with Ansible. There are other things we will be doing Concourse, such as building AMIs and Docker images etc, that are probably closer to the standard use-cases. 

So far this week, we have got the telia-oss `github-pr-resource` setup for validating our PRs, along with pipelines that use static inputs and we're loving it so far. It's quite different from Jenkins, but so far in good ways. Being able to update a pipeline using fly without having to first commit a `Jenkinsfile` and trigger a job you know will fail just so it can pick up the updated `Jenkinsfile` is awesome.  We're about three to fours weeks into the POC project.

**Comment by pieterza** [comment]
> A CI system that doesn't support parameterised builds!? We are currently running a POC for migrating to Concourse from Jenkins and I don't want to have to create _n_ versions of a job to cater for _n_ parameter permutations...

Stay on Jenkins.  We've gone full circle, as you'll see by my nearly 2 year old comment on this very thread.

**Comment by bukowa** [comment]
> Concourse is designed for, which is to be _continuous_, driven by a declarative configuration that runs in the background, triggered by external systems that act as the source of truth for all inputs.

What if the truth (the vars) are declared in the pipeline file and we can choose from them before the pipeline runs?

**Comment by obashistu** [comment]
This decision kills any hopes of using Concourse in place of Jenkins for us. 

I appreciate standing firm on project objectives though in this instance it seems a little petulant and moralistic. The decision to not implement useful features such as triggers and parametrised jobs leaves this software in a no-mans land as far as Im concerned - neither a good fit for one thing nor another. 

The concourse project may remain pure in its ambition but that means little when so many users have come and gone.

**Comment by IAXES** [comment]
Are there any other issues/tickets requesting this feature, or is it safe to say it's not going to happen?
